### PR TITLE
Respect lambda fxn errors and treat as 504

### PR DESCRIPTION
**Pull request by nfuden**
On 200 response with fxn error such as with timeouts treat it as a 504 and pass the 504 along to the caller from downstream.

### CODE REVIEW

File: changelog/v1.19.0-patch4/bump-envoy.yaml [hunk 1]
```diff
@@ -0,0 +1,6 @@
+changelog:
+- type: NEW_FEATURE
```
Comment (kdorosh) [inline changelog/v1.19.0-patch4/bump-envoy.yaml:2]:
```suggestion
- type: FIX
```

File: source/extensions/filters/http/aws_lambda/aws_lambda_filter.cc [hunk 2]
```diff
@@ -120,6 +121,28 @@ AWSLambdaFilter::decodeHeaders(Http::RequestHeaderMap &headers,
...
+
+Http::FilterDataStatus AWSLambdaFilter::encodeData(Buffer::Instance &,
+                                                        bool ) {
+  return Http::FilterDataStatus::Continue;
+}
+
+Http::FilterTrailersStatus
+AWSLambdaFilter::encodeTrailers(Http::ResponseTrailerMap &) {
+  return Http::FilterTrailersStatus::StopIteration;
+}
```
Comment (kdorosh) [inline source/extensions/filters/http/aws_lambda/aws_lambda_filter.cc:144]: why not move these to the header file like we did with `encode100ContinueHeaders`?

Comment (nfuden) [reply source/extensions/filters/http/aws_lambda/aws_lambda_filter.cc:144]: Yah makes sense. Do we have a general guideline for complexity allowed in header? Something like if it is anything more than a no-op then dont put it in header?

Namely I am even wondering if even the encode headers should go in to the header file.
My gut feeling is no but while we are having the discussion its good know.

Comment (kdorosh) [reply source/extensions/filters/http/aws_lambda/aws_lambda_filter.cc]: We have generally been following upstream envoy for conventions.

I was just calling out that we were being inconsistent with our approach here, but i don't have a super strong opinion either way

**Comment by kdorosh** [comment]
/skip-changelog changelog-bot bug can't discern between two possible valid next versions

File: source/extensions/filters/http/aws_lambda/aws_lambda_filter.cc [hunk 3]
```diff
@@ -120,6 +121,18 @@ AWSLambdaFilter::decodeHeaders(Http::RequestHeaderMap &headers,
   return Http::FilterHeadersStatus::StopIteration;
 }
 
+
+Http::FilterHeadersStatus 
+AWSLambdaFilter::encodeHeaders(Http::ResponseHeaderMap &headers, bool ) {
+
+  if (!headers.get(AWSLambdaHeaderNames::get().FunctionError).empty()){
+    // We treat upstream errors as if it was any other upstream error
```
Comment (EItanya) [inline source/extensions/filters/http/aws_lambda/aws_lambda_filter.cc:129]:
```suggestion
    // We treat upstream function errors as if it was any other upstream error
```

File: test/extensions/filters/http/aws_lambda/aws_lambda_filter_test.cc [hunk 4]
```diff
@@ -402,6 +402,22 @@ TEST_F(AWSLambdaFilterTest, NoCredsAvailable) {
...
+  Http::TestResponseHeaderMapImpl response_headers{
+    {"content-type", "test"},
+    {":method", "GET"},
+    {":authority", "www.solo.io"},
+    {":status", "200"},
+    {"x-amz-function-error", "fakerr"},
+    {":path", "/path"}};
+  auto res = filter_->encodeHeaders(response_headers, true);
+  EXPECT_EQ(Http::FilterHeadersStatus::Continue, res);
+  EXPECT_EQ(response_headers.get_(":status"), "504");
```
Comment (kdorosh) [inline test/extensions/filters/http/aws_lambda/aws_lambda_filter_test.cc:417]: nit, we can use `getStatusValue()` here

File: source/extensions/filters/http/aws_lambda/aws_lambda_filter.cc [hunk 5]
```diff
@@ -223,7 +236,7 @@ AWSLambdaFilter::decodeTrailers(Http::RequestTrailerMap &) {
 }
 
 void AWSLambdaFilter::lambdafy() {
-
+
```
Comment (kdorosh) [inline source/extensions/filters/http/aws_lambda/aws_lambda_filter.cc:239]: super nit on removing these whitespaces

File: source/extensions/filters/http/aws_lambda/aws_lambda_filter.cc [hunk 6]
```diff
@@ -120,6 +121,18 @@ AWSLambdaFilter::decodeHeaders(Http::RequestHeaderMap &headers,
...
+Http::FilterHeadersStatus 
+AWSLambdaFilter::encodeHeaders(Http::ResponseHeaderMap &headers, bool ) {
+
+  if (!headers.get(AWSLambdaHeaderNames::get().FunctionError).empty()){
+    // We treat upstream errors as if it was any other upstream error
+    headers.setStatus(504);
+    return Http::FilterHeadersStatus::Continue;
+  }
+  return Http::FilterHeadersStatus::StopIteration;
```
Comment (EItanya) [inline source/extensions/filters/http/aws_lambda/aws_lambda_filter.cc:133]: Why are we stopping iteration here?

Comment (kdorosh) [reply source/extensions/filters/http/aws_lambda/aws_lambda_filter.cc:133]: i think we should always return continue here`

Comment (nfuden) [reply source/extensions/filters/http/aws_lambda/aws_lambda_filter.cc:133]: Ah I thought I changed that good catch.

Comment (kdorosh) [reply source/extensions/filters/http/aws_lambda/aws_lambda_filter.cc:133]: i think we should always continue here

File: source/extensions/filters/http/aws_lambda/aws_lambda_filter.cc [hunk 7]
```diff
@@ -120,6 +121,18 @@ AWSLambdaFilter::decodeHeaders(Http::RequestHeaderMap &headers,
...
 }
 
+
+Http::FilterHeadersStatus 
+AWSLambdaFilter::encodeHeaders(Http::ResponseHeaderMap &headers, bool ) {
+
+  if (!headers.get(AWSLambdaHeaderNames::get().FunctionError).empty()){
+    // We treat upstream function errors as if it was any other upstream error
+    headers.setStatus(504);
+    return Http::FilterHeadersStatus::Continue;
```
Comment (kdorosh) [inline source/extensions/filters/http/aws_lambda/aws_lambda_filter.cc:131]: super nit, we can remove this line

File: source/extensions/filters/http/aws_lambda/aws_lambda_filter.cc (hunk 6 above)
Comment (nfuden) [reply source/extensions/filters/http/aws_lambda/aws_lambda_filter.cc]: Updated to always send continue and have that continue out of the if.

File: test/extensions/filters/http/aws_lambda/aws_lambda_filter_test.cc (hunk 4 above)
Comment (nfuden) [reply test/extensions/filters/http/aws_lambda/aws_lambda_filter_test.cc]: Updated per request. Definitely shows I grabbed an older example extension for cribbing the test from.

**Review by kdorosh — APPROVED** [review]

**Pull request merged**